File: benchmarks/python/tcp_async_bench.py
#!/usr/bin/env python3
"""Python TCP Async Benchmark (asyncio)
Proper async/await pattern matching TML async"""

import socket
import asyncio
import time

async def bench_async_tcp():
    """Async TCP bind benchmark using asyncio"""
    print("\n================================================================")
    print("  Python TCP Benchmarks: Async (asyncio)")
    print("================================================================\n")

    print("=== ASYNC TCP (asyncio - 50 concurrent) ===")
    print("  Binding to 127.0.0.1:0 (50 concurrent binds)\n")

    n = 50
    success = 0

    async def bind_async():
        """Async bind operation"""
        try:
            # Create socket
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

            # Non-blocking bind
            sock.bind(("127.0.0.1", 0))
            sock.close()
            return True
        except Exception:
            return False

    start = time.perf_counter()

    # Run all binds concurrently via asyncio gather
    tasks = [bind_async() for _ in range(n)]
    results = await asyncio.gather(*tasks)
    success = sum(results)

    elapsed = time.perf_counter() - start
    ns_elapsed = int(elapsed * 1_000_000_000)

    print(f"    Iterations: {n}")
    print(f"    Total time: {ns_elapsed // 1_000_000} ms")
    print(f"    Per op:     {ns_elapsed // n} ns")
    print(f"    Ops/sec:    {n * 1_000_000_000 // ns_elapsed if ns_elapsed > 0 else 0}")
    print(f"    Successful: {success}/{n}\n")

    print("================================================================\n")

if __name__ == "__main__":
    asyncio.run(bench_async_tcp())

File: benchmarks/node/tcp_async_bench.js
#!/usr/bin/env node
/**
 * Node.js TCP Async Benchmark (proper Promise.all async pattern)
 */

const net = require('net');

console.log('\n================================================================');
console.log('  Node.js TCP Benchmarks: Async (Promise.all)');
console.log('================================================================\n');

async function benchAsync() {
  console.log('=== ASYNC TCP (Promise.all - 50 concurrent) ===');
  console.log('  Binding to 127.0.0.1:0 (50 concurrent binds)\n');

  const n = 50;
  const start = process.hrtime.bigint();
  let success = 0;

  // Create 50 concurrent bind operations
  const promises = Array(n).fill().map(() => {
    return new Promise(resolve => {
      const server = net.createServer();

      server.listen(0, '127.0.0.1', () => {
        success++;
        server.close(resolve);
      });

      server.on('error', resolve);

      // Timeout after 5 seconds
      setTimeout(() => {
        server.close(resolve);
      }, 5000);
    });
  });

  // Wait for all to complete
  await Promise.all(promises);

  const end = process.hrtime.bigint();
  const nsElapsed = end - start;

  console.log(`    Iterations: ${n}`);
  console.log(`    Total time: ${Number(nsElapsed / 1_000_000n)} ms`);
  console.log(`    Per op:     ${Number(nsElapsed / BigInt(n))} ns`);
  console.log(`    Ops/sec:    ${nsElapsed > 0n ? Number((BigInt(n) * 1_000_000_000n) / nsElapsed) : 0}`);
  console.log(`    Successful: ${success}/${n}\n`);

  console.log('================================================================\n');
}

benchAsync().catch(console.error);
